### DISCUSSION TITLE
Konfiguration MQTT innerhalb open3e

### DISCUSSION THREAD

**Discussion by FRMB6** [open3e/open3e#159] [Q&A] [answered]
Hallo zusammen,

Deutsch scheint die Sprache der Meisten hier zu sein. Ich habe ein VX3 System mit zwei Wechselrichtern und 10 KWh Batterie, wobei derzeit der WR mit Batterie mit dem Pi verbunden ist.

Ich hänge derzeit etwas an der Übergabe der Messwerte vom Bus über MQTT an HomeAssistant. Ich muss dazu sagen, dass ich Kommandozeilenbefehle zwar gut nachvollziehen kann, jedoch nicht viel mehr.

Ich habe open3e auf einem frisch dafür aufgesetzten Pi zum Laufen gebracht, mittels

` open3e_depictSystem`

wurden viele DIDs gefunden und auf dem Pi (scheinbar in das Home Verzeichnis?) folgende Dateien geschrieben:

- devices.json
- Open3Edatapoints_XYZ.py

Von letzterer Datei 9 Stück mit den jeweiligen ECUs. Es wurden keinerlei "virtdata_XYZ" Files geschrieben.

Wie kann ich jetzt von den vielen DIDs die relevanten raussuchen und die regelmäßig mittels MQTT übergeben?

Vielen Dank schon einmal für eure Hilfe.

**Comment by Philip-Wiege** [comment]
Hi @FRMB6! Welche Datenpunkte / DIDs für dich relevant sind, kann ich dir nicht sagen. Ich denke anhand der Namen könntest du aber bestimmt schonmal SOC, PV und Wechselrichterleistung etc. identifizieren. Vielleicht kann die @DpunktS ja helfen. Ich glaube er hat auch ne VX3?

Ich habe bei mir open3e so laufen, dass die Argumente aus meiner Args-Datei verwendet werden. Dort habe ich auch meine für mich relevanten DIDs definiert. In der Args-Datei musst du natürlich noch die für dich passenden DID Nummern eintragen, sowie IP, Port, Username, Passwort, und die MQTT-Topics für senden und empfangen eintragen. Die Args-Datei liegt im open3e Ordner und wenn man open3e über Kommandozeile startet, sollte auf dem MQTT Server alles ankommen:

`open3e @args`

[args.txt](https://github.com/user-attachments/files/17948573/args.txt)

**Reply by FRMB6** [reply]
Hi Philip,

Einen Fehler meinerseits habe ich schon gefunden, ich hatte kein Unterverzeichnis open3e angelegt, also die Installation aus Versehen etwas abgekürzt. Deswegen wurde über depictsystem alles auf das Heimverzeichnis geschrieben. Schauen wir mal ob dann auch die "virtdata_XYZ" Dateien geschrieben werden, weil da müssten ja die Namen der DIDs drinstehen (?). Ich melde mich wenn ich das System nochmal gescannt habe.

**Reply by FRMB6** [reply]
Systemscan ist durchgeführt, nur wo habe ich jetzt eine Liste mit den DIDs, wo ich mir die für mich relevanten raussuchen kann? Ich meinte gelesen zu haben, dass diese noch in extra "virtdata_XYZ" Dateien geschrieben werden. Ansonsten habe ich im Terminal bereits korrekt aussehende DID Namen gesehen.

**Reply by Philip-Wiege** [reply]
Wenn ich die DIDs nachschaue, gucke ich einfach in die Quelldatei auf Github: https://github.com/open3e/open3e/blob/master/src/open3e/Open3Edatapoints.py

Das mit Virtdata ist glaube ich ein Feature, welches @philippoo66 kennt? Ich glaube wenn man Open3e_depictSystem mit dem Parameter -s ausführt, werden zu jedem DID auch die echten, gelesenen Werte deiner Anlage in die Virtdata Files gespeichert. Das ist zum Betrieb von Open3E erstmal nicht notwendig aber zur Entwicklung / Simulation hilfreich.

**Reply by philippoo66** [reply]
@FRMB6 
>  nur wo habe ich jetzt eine Liste mit den DID

die datapoints_6xy.py die der Systemscan erstellt stehen meiner Info nach und soweit ich erinnere in dem Verzeichnis, in dem du das depictSystem ausgeführt hast (ich weiss seit pip auch nicht mehr sooo sicher Bescheid, aber ich denke, der Ort passt).

Die sind allerdings etwas 'kryptisch', weil es nur für deine Anlage zutreffende Overlays zu der generischen datapoints.py sind. Details wie das funktioniert und was es bedeutet kannst du hier https://github.com/open3e/open3e/wiki/035-Open3Edatapoints-%E2%80%90-die-Zusammenh%C3%A4nge nachlesen.

**Reply by philippoo66** [reply]
ps. die virtdata Files sind tatsächlich die Daten für die Simulation [virtual3e](https://github.com/philippoo66/vitualE3) oder [hier](https://github.com/MyHomeMyData/vitualE3) der fork von Jürgen. 

wenn du uns einen Gefallen tun willst / unsere Entwicklungsarbeit unterstützen, dann schickst du diese Dateien zusammen mit den datapoints_6xy und der config.json an open3e/at/web/Punkt/de und schreibst noch kurz ein paar Worte zu deiner Anlage dazu (Komponenten, ...) 

Achtung: die virtdata nicht irgendwo öffentlich hochladen! weil da Daten drin sind / sein können, die bei ungenehmigter Veröffentlichung Verstoss gegen die Datenschutzgrundverordnung darstellen (bsw. die Kontaktdaten deines Installateurs und seine Viessmann-ID oder ggf auch die geografischen Daten deiner Anlage oder deine Adresse und sowas).

**Reply by FRMB6** [reply]
Alles klar, das hat mir jetzt schonmal total weitergeholfen! Bevor ich mich mit MQTT beschäftige wollte ich gerne zuerst ein Proof of Concept, i.e., zu sehen dass der Pi wirklich die Daten vom WR abgreift. Das funktioniert nun.

Mittels

`open3e -c can0 -dev vx3 -r 1824 -v`

erhalte ich nun

`vx3 1824 ThreePhaseInverterCurrentPower {"cumulated": 321.0, "L1": 100.0, "L2": 119.0, "L3": 104.0}`

als Antwort. Als nächstes steht nun an, eine Liste mit DIDs zu erstellen. Wie ist hier die best practise? Die Liste unter https://github.com/open3e/open3e/blob/master/src/open3e/Open3Edatapoints.py habe ich, daher habe ich auch die DIDs mir rausgesucht.. das ist allerdings ohne zu Wissen wonach man sucht etwas aufwendig. Gibt es eine Liste mit relevanten DIDs für einen VX3?

Vielen  Dank schonmal dass ihr mich hier etwas "an die Hand nehmt"!

**Reply by philippoo66** [reply]
moin @FRMB6 !

du kannst als ersten Schritt mal schauen, welche Datenpunkte überhaupt auf deiner VX3 existieren. Je nachdem ob die VX3 die Main ECU ist oder eine 'Slave' ECU (z.B. wenn die VX3 'hinter' einer Vitocal hängt) findest du die Datenpunkte in Open3Edatapoints_680.py oder in Open3Edatapoints_6a1.py. 

Nicht wundern, da stehen viele 'None's drin - das bedeutet dann, dass dieser Datenpunkt (zumindest gemäß seiner Länge) dem in Open3Edatapoints.py benannten Codec folgt. Details zum Overlay System findest du [hier](https://github.com/open3e/open3e/wiki/035-Open3Edatapoints-%E2%80%90-die-Zusammenh%C3%A4nge).

Dann gibt es noch die Batteriemodule, die ggf. unter Open3Edatapoints_6cc/6cd/... .py auftauchen, aber ich glaube nur, wenn du am Internen CAN Bus hängst. Schau dir mal die devices.json an, welche ECUs da so drin stehen.

Die virtdata_6xy.txt werden übrigens nur geschrieben, wenn du bei depictSystem die Option -s mit aufrufst. -s steht für 'Simulation', und dementsprechend enthalten die Dateien keine Namen oder sowas, sondern die Datenpunkt-Daten in hex Form. Damit kannst du (oder wir wenn du uns den Datensatz zur Verfügung stellst) die Anlage simulieren basierend auf dem Zustand zum Zeitpunkt des Snapshot. Das funktioniert mit der [virtualE3](https://github.com/philippoo66/vitualE3).

**Comment by MyHomeMyData** [comment]
@FRMB6 Infos zu VX3 Datenpunkten gibt's z.B. hier:
https://github.com/open3e/open3e/discussions/95

**Comment by FRMB6** [comment]
Moin, dank des files https://github.com/open3e/open3e/blob/master/src/open3e/Open3Edatapoints.py und dem Beitrag von @diefraschw (https://github.com/open3e/open3e/discussions/95) habe ich meine Liste mit DIDs soweit zusammen. Bei Bedarf kann diese ja noch erweitert werden!

Nun zu MQTT, ich brauche natürlich mal wieder ein bisschen Hilfe. Ich möchte die Daten gerne an Home Assistant übergeben, dort läuft auch schon die MQTT Integration, die zuverlässig Werte meiner Buderus Wärmepumpe empfängt (Wildwuchs, ich weiß..).

Ich habe jetzt folgenden Aufruf gebastelt, in den nach Bedarf ja noch mehr DIDs reingepackt werden können:

`open3e -c can0 -dev vx3 -r 1831 -v -m IPHomeAssistant:1883:open3e -mfstr "{didNumber}_{didName}"`

Dies gibt folgende Antwort:

```
mqtt broker disconnected. reason_code = Unspecified error
vx3 1831 PhotovoltaicCurrentStringPower {"String1": 0.0, "String2": 0.0, "String3": 0.0}
closing 0x680 - bye!
```

Könnt ihr mir nochmal auf die Sprünge helfen? Besten Dank allen Unterstützern hier!

**Reply by MyHomeMyData** [reply]
Versuche mal ein
```
mosquitto_pub -d -h IPHomeAssistant -t open3e/test -m "hello world"
```

Vielleicht helfen die Debug-Meldungen weiter.

**Reply by FRMB6** [reply]
`-bash: mosquitto_pub: command not found`

scheine ich wohl noch nicht installiert zu haben?

**Reply by MyHomeMyData** [reply]
```
sudo apt install mosquitto mosquitto-clients
```

**Reply by FRMB6** [reply]
Ist installiert. 

Oben genannter Befehlte ergibt nun:

`Unable to connect (Lookup error.).`

**Reply by MyHomeMyData** [reply]
D.h. Dein Raspi kann "IPHomeAssistant" nicht erreichen. Ein `ping IPHomeAssistant` dürfte auch nicht funktionieren.

Das ist dann ein Netzwerkthema und hat mit open3e nichts zu tun.

**Reply by MyHomeMyData** [reply]
Wo läuft denn der MQTT-Broker? Docker, virtuelle Maschine, ...?

**Reply by FRMB6** [reply]
Also. Mein Fehler, bei

`mosquitto_pub -d -h IPHomeAssistant -t open3e/test -m "hello world"`

hatte ich noch den Port an die IP Adresse angehängt. Wenn ich das nicht mache kommt ein

`Client (null) sending CONNECT
Client (null) received CONNACK (5)
Connection error: Connection Refused: not authorised.
Error: The connection was refused.`

zurück. Ich vermute weil ich kein Nutzer und/oder Passwort eingegeben habe?

Home Assistant läuft nativ auf einem Pi, mit dem MQTT Plugin.

**Reply by MyHomeMyData** [reply, marked as answer]
Ja, "not authorised" dürfte bedeuten, dass Du User und Passwort angeben musst. Das hast Du ja im Plugin festgelegt, oder? Eine andere Möglichkeit ist, dass der Broker gar keine Verbindungen von anderen Rechnern annimmt. Auch das dürfte Teil der Plugin-Konfig sein.

**Reply by FRMB6** [reply]
So, alles abgeglichen, keine Fehlermeldung mehr.. DANKE!! Fehler lag.. wie sollte es anders sein.. bei mir.

**Comment by floh2307** [comment]
Hi Philipp,

ich hätte eine Frage bzgl des MQTT Scrips im Open3e.

Beim Optolink Splitter habe ich die LWT Message also (online oder offline)
Gibts die Möglichkeit für open3e und E3onCan auch? Bei diesen 2 Skripten hab ich nämlich diese Message nicht.

Gruß
Florian

**Reply by philippoo66** [reply]
hai Florian!

funktioniert das bei dir gut? bei mir zeigt es immer das Gegenteil an hab ich das Gefühl... 

Wenn du magst, bauen wir das beim nächsten Update mit ein. Du kannst es aber bis dahin auch einfach selber rüberkopieren. Ich hab im Splitter den MQTT Part aus open3e übernommen, von daher kannst du das LWT einfach aus dem Splitter in's open3e kopieren. Es sind nur 3 Zeilen:

![grafik](https://github.com/user-attachments/assets/c7490241-e778-40a0-919c-f2a75541cc02)

![grafik](https://github.com/user-attachments/assets/c27898b0-d667-456a-8031-27ffbdef441d)

bei open3e ist der mqttTopic aber `mqttTopic` und nicht `settings_ini.mqtt_topic` wie beim Splitter. das müsstest du ersetzen.

**Reply by floh2307** [reply]
Hi Philipp,
ja beim Splitter und bei den anderen MQTT geräten geht es mit dem LWT eigentlich gut.

Irgendwie steh ich grade etwas auf dem schlau. Wo muss das genau beim open3e rein? in die args?

**Reply by philippoo66** [reply]
ne, in den Code ;-)

ich habe dir einen neuen branch gemacht: [feat_lwt](https://github.com/open3e/open3e/tree/feat_lwt) Da habe ich die 3 Zeilen eingebaut. VS Code mäkelt die Variable mqttTopic in OnConnect und in OnDisconnect nicht an, obwohl ich mir nicht ganz sicher bin, ob es die (richtig) kennt. 

könntest du das mal ausprobieren und berichten?!

du muss dafür nur das Modul Open3Eclient.py auf deinem PI durch das aus dem feature branch ersetzen und open3e neu starten. (am besten das org. Modul vorher umbenennen, damit du das Ganze bei Misserfolg problemlos wieder rückgängig machen kannst)

danach sollte im MQTT Baum unter open3e (wichtig! oder unter den 'Stamm', den du mit `-m: <ip>:1883:<Stamm>` angegeben hast) der leave "LWT" mit dem Wert 'online' oder 'offline' erscheinen.

**Reply by philippoo66** [reply]
ps. wenn du ein pip install gemacht hast, steht open3e ja irgendwo nicht im Benutzerverzeichnis. dann musst du die Dateioperationen mit sudo machen.

**Reply by philippoo66** [reply]
pps: ich habe es grad ausprobiert - bei mir funktioniert es :-)

**Reply by floh2307** [reply]
Hab eben mal die Datei reinkopiert (src/open3e/Open3Eclient.py)), komplett neu gestartet,  aber LWT kommt irgendwie nicht im Broker an.

![image](https://github.com/user-attachments/assets/83409875-cfc7-470c-9f81-c79e97ebf818)

**Reply by philippoo66** [reply]
hattest du ursprünglich ein pip install gemacht? dann stehen die Module, die benutzt werden, woanders (unter /etc oder /opt oder so. guck da noch mal.

ich hab grad noch ein "hello LWT" eingebaut, das muss beim Start ausgegeben werden:

![grafik](https://github.com/user-attachments/assets/573bb169-f0af-4dc6-a27a-c3d8448cf4dd)

damit bekommst du

![grafik](https://github.com/user-attachments/assets/1fb5c680-5ceb-4cf2-92df-b076a087825d)

wenn das 'hello LWT' nicht ausgegeben wird, hast du das falsche Open3Eclient.py ersetzt ;-)

**Reply by philippoo66** [reply]
ChatGPT:

Wenn du mit pip install ein Python-Modul auf einem Raspberry Pi installierst, wird das Modul standardmäßig in einem bestimmten Verzeichnis auf deinem System installiert. Wo genau, hängt davon ab, wie du pip verwendest und welches Python-Environment du aktiv hast. Hier sind die gängigsten Fälle:

    Systemweite Installation (global): Wenn du pip install ohne ein virtuelles Environment und ohne Administratorrechte (also ohne sudo) verwendest, wird das Modul in das Verzeichnis für die Python-Pakete des aktuellen Benutzers installiert:
        Python 3: In der Regel wird es in einem Verzeichnis wie /home/benutzername/.local/lib/python3.x/site-packages/ installiert.

    Installation mit sudo (systemweit): Wenn du sudo pip install verwendest, wird das Modul systemweit für das Python-System installiert, und zwar in einem Verzeichnis wie:
        /usr/local/lib/python3.x/dist-packages/ oder /usr/lib/python3.x/dist-packages/ (je nach Distribution und Python-Version).

    Virtuelles Environment (empfohlen): Wenn du ein virtuelles Environment (z. B. mit python3 -m venv venv) nutzt, wird das Modul innerhalb dieses Environments installiert:
        Das Modul wird dann in einem Ordner wie venv/lib/python3.x/site-packages/ installiert, wobei venv der Name des virtuellen Environments ist.

Um herauszufinden, wo ein Modul installiert wurde, kannst du das folgende Kommando verwenden, nachdem du das Modul installiert hast:

python3 -m site

Dies gibt dir Details zu den Standardverzeichnissen, in denen Python nach Modulen sucht.

Zusätzlich kannst du auch das Installationsverzeichnis eines spezifischen Moduls mit diesem Befehl herausfinden:

pip show <modulname>

Dies zeigt dir die Installationsinformationen des jeweiligen Moduls, einschließlich des Installationspfads.

**Reply by philippoo66** [reply]
mach mal ein `pip show open3e` oder sowas

**Reply by floh2307** [reply]
Jetzt hab ich den richtigen Ordner gefunden und jetzt geht es auch :-)

**Comment by mattokoester** [comment]
Hallo,
erstmal einen riesen Dank an alle, die hier so krass viel leisten!!
Ich würde mich über Hilfe freuen...! Ich habe leider nur wenig Erfahrung mit der Thematik und habe mich da durchgebissen bis hier.

Grob die Konfiguration:
Vitocal 250 und E380CA, PI3 mit Canbus, PI4 mit HA

Was bislang läuft:
ich habe auf einem Pi3 einen RS485 Can Hat angeschlossen. Dieser liefert auch Daten aus dem Canbus. Auf dem PI3 habe ich dann Open3E installiert.
Auf dem PI4 in HA habe ich Mosquitto/ MQTT hinzugefügt. Für die Entitäten habe ich die Vorlage von Wolfgang verwendet. Das hat auch alles gut funktioniert. Leider kommen im entsprechenden Dashboard aber keine Werte an.
Ich würde mich riesig freuen, wenn einer da Abhilfe wüsste :)

VG Jonas

**Reply by jwende** [reply]
Eigentlich würde ich empfehlen, für ein neues Problem einen neuen Thread aufzumachen.
Kurze Abfolge für die Fehlersuche:
- du kannst mit einem externen Tool (MQTT Explorer zB) sehen, dass Daten im MQTT ankommen und der Topic Baum korrekt aussieht ?
- du kannst im HA die Daten der VC250 sehen ?
![image](https://github.com/user-attachments/assets/f2cb9698-6d03-4ad2-96ac-10ca79c0a070)
![image](https://github.com/user-attachments/assets/e4fad1be-e01c-4a1e-88fa-3a866afc217a)
Wenn das stimmt, muss man sich 'nur' noch mit der korrekten Zuordung im Diagramm auseinandersetzen.
![image](https://github.com/user-attachments/assets/3e9d8b44-d1fd-4c6c-b80a-25627d5ddc14)
![image](https://github.com/user-attachments/assets/1b64ce84-fbaf-4038-b669-117d51ad5b4b)

Ich kann es dir im Augenblick leider nicht einfacher machen. Sorry
J.

**Reply by mattokoester** [reply]
Hi,
die Verbindung scheint zu stehen.
Jedoch hat die VC250 nur ein Topic:
![Screenshot 2025-02-11 174200](https://github.com/user-attachments/assets/904c7499-15c5-4f69-a198-43ed879f56fc)

NACHTRAG: ich habe die Konfig. Datei in Open3E nochmal neu eingefügt. Jetzt flutscht alles!! Trotzdem Danke :)

**Reply by philippoo66** [reply]
_offline_ ist schon mal nicht gut, da sollte eigentlich _online_ stehen, wobei das bei mir auch öfter quer geht.

dazu ist zu beachten, dass das LWT retain ist, es kann also von vor 'Urzeiten' her stammen.

kannst du bitte mal deinen Aufruf von open3e posten bzw. die args Datei!? ich vermute hier liegt der Hase im Pfeffer...

**Reply by philippoo66** [reply]
wichtig ist erstmal 
`-m 192.168.178.109:1883:open3e` (Beispiel), wobei die IP die von dem Raspi mit dem MQTT Broker sein muss (der PI4?)

Falls du einen 'geschützen' Zugriff auf den MQTT Broker eigerichhtet hast, natürlich auch das
`-muser MQTTUSER:PASSW` natürlich mit deinem user und pwd

**Reply by mattokoester** [reply]
Ich hatte oben noch einen Nachtrag unter das Bild gesetzt: läuft jetzt alles! Die Konfig. Datei hatte irgendeinen Dreher. Ich habe sie einfach neu aufgesetzt.

**Reply by philippoo66** [reply]
> läuft jetzt alles!

das klingt gut!
